schemachangerccl: TestBackup_base_alter_table_add_check_unvalidated failed

On [this](https://teamcity.cockroachdb.com/viewLog.html?buildId=8816895&buildTypeId=Cockroach_BazelEssentialCi) on probably an unrelated change:
```
Failed
=== RUN   TestBackup_base_alter_table_add_check_unvalidated/backup/restore_stage_0_of_0
    cumulative.go:1145: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/10765/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/schemachangerccl/schemachangerccl_test_/schemachangerccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/schemachangerccl/cumulative.go:1145
        	            				/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/10765/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/schemachangerccl/schemachangerccl_test_/schemachangerccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/schemachangerccl/cumulative.go:1304
        	Error:      	Received unexpected error:
        	            	pq: internal error: executing declarative schema change PreCommitPhase stage 1 of 2 with 1 MutationType op (rollback=false) for ALTER TABLE: close of closed channel
        	Test:       	TestBackup_base_alter_table_add_check_unvalidated/backup/restore_stage_0_of_0
    --- FAIL: TestBackup_base_alter_table_add_check_unvalidated/backup/restore_stage_0_of_0 (4.70s)
```

Jira issue: CRDB-24774

--- a/pkg/sql/schemachanger/sctest/cumulative.go
+++ b/pkg/sql/schemachanger/sctest/cumulative.go
@@ -995,11 +995,21 @@ func Backup(t *testing.T, path string, newCluster NewClusterFunc) {
 		}
 
 		stageChan := make(chan stage)
+		var closedStageChan bool // mark when stageChan is closed in the callback
 		ctx, cancel := context.WithCancel(context.Background())
 		_, db, cleanup := newCluster(t, &scexec.TestingKnobs{
 			BeforeStage: func(p scplan.Plan, stageIdx int) error {
+
+				// If the plan has no post-commit stages, we'll close the
+				// stageChan eagerly.
 				if p.Stages[len(p.Stages)-1].Phase < scop.PostCommitPhase {
-					if stageChan != nil {
+
+					// The other test goroutine will set stageChan to nil later on.
+					// We only want to close it once, and then we don't want to look
+					// at it again. Avoid the racy access to stageChan by consulting
+					// the bool.
+					if !closedStageChan {
+						closedStageChan = true
 						close(stageChan)
 					}
 					return nil

